src/stream/src/executor/source/source_backfill_executor.rs

```diff
@@ -665,11 +648,25 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
                                         .await?;
                                     }
 
+                                    if !state_table_initialized {
+                                        // This is for self.backfill_finished() to be safe: wait until this actor can read all actors' written data.
+                                        // We wait for 2nd epoch
+                                        let epoch = barrier_epoch.prev;
+                                        tracing::info!("waiting for epoch: {}", epoch);
+                                        state_store
+                                            .try_wait_epoch(
+                                                HummockReadEpoch::Committed(epoch),
```

Comment:
Currently we may have checkpoint and non-checkpoint barrier. When we have checkpoint frequency more than 1, it's likely that the second barrier is a non-checkpoint barrier. 

Let's say the first epoch is 1, and then the second epoch is 2, which is a non-checkpoint epoch, and then next checkpoint epoch is 5. Here the `try_wait_epoch` on epoch 2 will finish only when the epoch 5 has finished the checkpoint. However, we in-place block the whole executor, which will block handling barrier 3, 4, 5, and further block the checkpoint of epoch 5, and then cause deadlock.

The correct implementation should be waiting on the next checkpoint epoch.